### DISCUSSION TITLE
Corpus analysis

### DISCUSSION THREAD

**Discussion by iandoug** [binarybottle/engram-es#21] [Ideas]
Hi

Okay, the Leipzig corpus is better if not perfect.

Attached analysis of  character frequency in Leipzig and Gutenberg, plus average. The data scientists can advise if that is the correct methodology :-)
It's sorted by character rather than count.

[spanish-character-frequency-v1.ods](https://github.com/binarybottle/engram-es/files/6963869/spanish-character-frequency-v1.ods)

Final Leipzig corpus : 
~/1web/spanish/leipzig $ wc spanish_leipzig-clean2.txt 
   5,480,692 lines,   265,281,891 words,  1,610,201,701 characters

Bigrams coming.

**Comment by iandoug** [comment]
Bigrams. Sorted. All permutations.

Do you need trigrams?

[spanish-bigram-frequency-v1.ods](https://github.com/binarybottle/engram-es/files/6964380/spanish-bigram-frequency-v1.ods)

**Reply by binarybottle** [reply]
Fantastic!  I don't need trigrams.

**Reply by iandoug** [reply]
I think Den (BEAKL) is going to have a go with Opt and trigrams will be useful. Will see tomorrow.

**Comment by iandoug** [comment]
First attempt at Cervantes Monkey. Needed to fiddle a bit to get Unicode to behave.

100k text and frequencies attached. Will try 1MB overnight ... takes about 4+ hours if I remember correctly.

[monkey-freq.txt](https://github.com/binarybottle/engram-es/files/6971370/monkey-freq.txt)
[spanishmonkeytest.txt](https://github.com/binarybottle/engram-es/files/6971371/spanishmonkeytest.txt)

**Reply by iandoug** [reply]
Memory a bit rusty ... the 1MB version takes over a day.

Video of it at work. The "size" in the yellow text is the number of bigrams left to process. It speeds up as that gets smaller, to about 15 characters per second. The stack started an about 1.2 MB (1MB + 20% spare)

Sorry it's a mkv in zip, Github does not allow mkv and VLC is not co-operating to convert it... some encoder not found.

[cervantes-monkey-2021-08-12_11.07.05.zip](https://github.com/binarybottle/engram-es/files/6974237/cervantes-monkey-2021-08-12_11.07.05.zip)

**Reply by iandoug** [reply]
Decided to see how fast it ran on the new PC. 
Much faster... already at over 110 chars/second.
Screen recorder can actually save mp4...

https://user-images.githubusercontent.com/12673886/129176050-35398e4d-9bad-4129-86ad-d55de953c59b.mp4

**Comment by iandoug** [comment]
Trigrams plus counts from Leipzig corpus. 

I cut it off at the point where the count/highest is 0.01, so just the 1222 most common. Should be enough.

[spanish-trios-leipzig.csv](https://github.com/binarybottle/engram-es/files/6973713/spanish-trios-leipzig.csv)
tab-separated.

BTW, before importing into spreadsheet, highlight the first column and Format Cells to text, else the stuff may get interpreted as a formula. Spreadsheets are too clever for their own good. This applies to the other .csv frequency files as well.

Cheers, Ian

**Comment by iandoug** [comment]
1MB chained bigrams, plus char frequency.

Original job still running :-)

[cervantes-1mb.txt-freq.txt](https://github.com/binarybottle/engram-es/files/6974960/cervantes-1mb.txt-freq.txt)
[cervantes-1mb.txt](https://github.com/binarybottle/engram-es/files/6974961/cervantes-1mb.txt)

**Comment by iandoug** [comment]
I think that concludes things with the corpus.

Will add the texts to KLAnext, but need to revise the user interface a bit ...

**Reply by Lobo-Feroz** [reply]
Awesome work you did here, Ian,

I'm sure this will be insanely useful for future keyboard designers and testers.

We're very lucky to have you around !

**Reply by iandoug** [reply]
Other size chained bigrams and analysis.
Never going to get exact char frequency as in full corpus, but should be good enough for development use,

10, 20, 30, 40, 50, 60, 100k, 1MB
[spanish-chained-bigrams.zip](https://github.com/binarybottle/engram-es/files/6984856/spanish-chained-bigrams.zip)

**Comment by binarybottle** [comment]
Thank you so much, @iandoug!

Now that I've returned from my travels, I will run the engram-es layout optimization code on your cleaned-up version of the Leipzig corpus, which is the largest and I hope most representative of what people would type.

**Reply by iandoug** [reply]
Leipzig does not include "book type" text ... dialogue etc.

**Reply by binarybottle** [reply]
According to https://wortschatz.uni-leipzig.de/en/download/Spanish:

> Used text material was taken from news websites (typically on a daily basis via RSS feeds).
> 
> Used text material was crawled from news websites and may be older than the specified year.
> 
> Used text material was taken from randomly chosen Web sites.
> 
> Used text material was taken from Wikipedia dumps.

Did you grab all of these?  Even these have different volumes, so we have to be careful about bias.

**Reply by iandoug** [reply]
I took the largest file in each row, except for those labelled as South/Central American.
There were a lot of non-Spanish names etc , which got dropped.... I dropped the entire offending line.
See the stats above ... it's bigger than most corpora I've seen analysed.

Will look at bigrams tomorrow.

**Comment by binarybottle** [comment]
> Okay, the Leipzig corpus is better if not perfect.
> It's sorted by character rather than count.

@iandoug -- There appears to be a redundant letter "i":

> ¡ 80563
> i 139041230

**Reply by iandoug** [reply]
One is i. The other is inverted exclamation. Github needs better fonts.
i¡

**Reply by binarybottle** [reply]
¡Got it -- thanks!

**Comment by binarybottle** [comment]
@iandoug -- I would like to put together a text data repo for Spanish like the one I did for English, with the modified files and a readme with a description of your modifications -- could you provide both?  And, not speaking or reading Spanish, I don't know how representative each of these categories of text data may be for what people will type. Do you have any sense what the relative representativeness might be, so we can weight them accordingly? What I'm thinking is if you could provide the count of the one-grams and bigrams for each category (news, websites, etc.), then we can combine them with each other and with the Gutenberg counts as we see fit, and others can do so for themselves in the future.

**Reply by binarybottle** [reply]
I added the following to the README and the notebook:

> ### Factors used to compute the engram-es layout <a name="factors">
>   - Spanish letter and letter bigram frequency data from a cleaned-up version of the Leipzig Spanish corpus. The largest file was downloaded from each row of the [original version](https://wortschatz.uni-leipzig.de/en/download/Spanish), except for rows corresponding to Central or South American countries, and lines containing non-Spanish names and words were removed.
>

**Reply by iandoug** [reply]
>     * except for rows corresponding to Central or South American countrie

I see they list USA and Andorra, so rather say except for rows specified as not Spain.

**Reply by iandoug** [reply]
The modified files are quite large. The final "cleaned" (as best possible) is 1.5GB, down from 3.4 GB.

1. Extract all the "sentences" files from the archives.
2. Merge into 1 file. 
3. Cleanup.. This was cyclical, including stripping off the leading line numbers, fixing some words, dropping sentences with unknown words / junk, and re-writing into paragraphed text to make it more 'natural'.
4. The final 'bad words' file that led to lines being dropped is attached. Some of these are probably kosher Spanish, but only occur a few or 1 times, working through 1.6 million of them will take a lot of time. Final corpus had 265 million words.
5. Rewriting as paragraphs also reduced file size by dropping around 80% of the carriage returns.

[badwords.zip](https://github.com/binarybottle/engram-es/files/6992282/badwords.zip)

I can't upload 1.5GB here. Was planning on posting it on Zenodo.

If you want counts by category I would need to redo it all in Categories. Also, their first section has no category. From my experience with doing the English corpus, I don't think there are dramatic differences between different web sites, I hope they did not include social media sites because that would push up the Junk Level. 
But books vs Web, yes, definite differences.

**Reply by iandoug** [reply]
The only-uppercase-letters bigrams for Leipzig is running ... takes a while.

**Reply by binarybottle** [reply]
@iandoug -- zenodo is a great idea.  Alternatively or in addition I was thinking osf.io, since I use this free service and their max cap per file is around 5gb.  Let's place all files and a description in one or both, then link to it from our documentation.

**Reply by Lobo-Feroz** [reply]
> badwords.zip

Hey @iandoug , the majority of the words in this file seem valid. I don't know at this point if it would be useful, but I could clean this up.

**Reply by iandoug** [reply]
Thanks. :-)
There are 1.6 million words in that file.... at the rate of 1 a second,  it will take 444 hours == 18 * 24. (or 3 times that long if 8 hours a day)
A lot seem to be Proper Names.
I don't think an extra million words added to the existing 256 million will change the character or bigram frequency much, but if you're bored, don't let me stop you :-)
As Arno said, I think we're at the point of diminishing returns.

**Reply by Lobo-Feroz** [reply]
Hehe, well, I didn't mean all of them. Maybe only the ones above a threshold.

The thing is that, since almost all words are valid, I can just do a scan to find out if there's any odd sticking out. Way faster than 1 per second. 

In badwords.txt, starting at "tomarla 765" (valid), these are the non-valid words:

- trav	761 (possibly "través")
- relaci	653 (possibly "relación")
- aacute	637 (html)
- iacute	623 (html)
- todav	622 (possibly "todavía")
- ltima	578 (última)
- atenci	551 (atención)
- decisin	547 (decisión)
- difcil	530 (difícil)
- obligatorioBorrar	517 ("obligatorio", "Borrar")
- poltico	509 (político)
- xito	476 (éxito)
- decisi	475 (decisión)
- posici	474 (posición)
- afirmou	461 (???)
- pelcula	454 (película)
- cuesti	452 (cuestión)
- declar	427 (declaró)
- elEconomista	422 (el Economista)
- revisin	418 (revisión)
- razn	405 (razón)

That would be all down unto "tomarnos 399" (valid), which would mean all words with 400+ instances. Didn't take more than a few minutes.

**Reply by iandoug** [reply]
Thanks. I can fix the the file, will then need to run various programs to rewrite and recompute.
Weekend project :-)

Cheers, Ian

**Reply by Lobo-Feroz** [reply]
@iandoug if you want I can make a run for the majority of that file. Say, all words with 100+ instances, and leave it at that. Won't take that long. I'll post it here hopefully this afternoon.

Out of curiosity, the actual "bad words", are you fixing them with these suggestions or just removing them?

**Reply by iandoug** [reply]
Have not started on these yet. For previous lists, I tried to fix, but only those surrounded by a space on either side, to avoid fixing strings in the middle of a word. That would have missed any next to punctuation. 
After that, I stripped all punctuation, checked the line for bad words, and dropped line if any found.

Realised just now I could undo the HTML entities as a first step.

If you can do down to a 100, that would be great, thanks :-)

cheers, Ian

**Reply by Lobo-Feroz** [reply]
@iandoug, review of badwords.txt from  "tomarnos 399" (valid), downwards to "recesiones 250" (valid).

These are invalid words:

```
llev 394 (llevó ???)
soluci 392 (solución)
és 388 (es) (without the accent)
eliminacin 381 (eliminación)
publicacin 381 (publicación)
fcil 377 (fácil)
ballottage 376 (french word)
muchisimo 373 (muchísimo)
kilmetros 370 (kilómetros)
mircoles 365 (miércoles)
ttulo 346 (título)
ás 344 (as) (without the accent)
vctimas 343 (víctimas)
restructuración 339 (reestructuración) (double 'e')
logr 338 (logró)
¨ 336 (special case: it's not a valid word, just a character)
fín 335 (fin) (without the accent)
acordo 335 (acordó)
creaci 334 (creación)
posicin 333 (posición)
expresi 330 (expresión)
oido 330 (oído)
realiz 321 (realizó)
intervenci 315 (intervención)
electrónicoCompartir 305 (electrónico Compartir) (two non-related words)
represi 303 (represión)
ocupaci 301 (ocupación)
compaeros 293 (compañeros)
presion 291 (presión)
opci 288 (french acronym)
reunen 288 (reúnen)
monton 281 (montón)
versin 272 (versión)
teor 271 (teoría ???)
últimas 271
solucin 266 (solución)
priísmo 265 (priismo)
televicentrotv 265 (televicentro tv)
condici 265 (condición)
convirti 265 (convirtió)
bién 264 (bien)
ocasin 262 (ocasión)
versi 262 (versión)
funci 260 (función)
eleccin 258 (elección)
fuenteaumentar 256 ("fuerte" "aumentar") (two unrelated words)
incluída 255 (incluida)
contraseñaRecordar 253 ("contraseña" "recordar")
```

One thing that clearly sticks out is that the majority of words here are perfectly valid, and some of them are common and should be in any dictionary. I don't know if the problem is that there are no actual good free/open source spanish dictionaries out there, if that's the case maybe we should use this opportunity to create a proper one.

Could there a way to have a similar file to this one, like goodwords.txt, having correct words and their counts? This could have several uses, for instance we could feed, say, the 10.000 more common words to a typing app like type-fu.com for training purposes.

**Reply by iandoug** [reply]
I got lists of words, but clearly incomplete.
I suppose part of the problem is any understanding of how Spanish changes words from one use case to the next.... whether tense, gender, whatever. So if every possibility was not covered in the word lists, then there will be gaps.

Last version from earlier this month, hope they are all good. Includes foreign proper names too... they also get used in real life.

Will have revised list tomorrow probably.

[goodwords2.zip](https://github.com/binarybottle/engram-es/files/7024171/goodwords2.zip)

**Reply by Lobo-Feroz** [reply]
> goodwords2.zip

Thanks Ian! This is great to have. Will try processing and uploading it to type-fu.

> I suppose part of the problem is any understanding of how Spanish changes words from one use case to the next.... whether tense, gender, whatever. So if every possibility was not covered in the word lists, then there will be gaps.

This is exactly what happens with the vast majority of the words in badwords.txt. Most of them are perfectly valid words, only with less common gender, plurals, or verb tense. Exactly this.

**Reply by iandoug** [reply]
Sorry I screwed up ... I uploaded the wrong bad words file. Was the original.
Should have been this one. How do these look?
[badwords2.txt](https://github.com/binarybottle/engram-es/files/7025500/badwords2.txt)

**Reply by Lobo-Feroz** [reply]
> Sorry I screwed up ... I uploaded the wrong bad words file. Was the original.
> Should have been this one. How do these look?
> [badwords2.txt](https://github.com/binarybottle/engram-es/files/7025500/badwords2.txt)

This one has much less words than the previous one, and most of them have typos from missing characters, missing the accute accent and so on. Most words here are actually invalid.

That means all valid words from the previous file have been already added to your dictionary? That's good news.

How could I help with this one?

**Reply by iandoug** [reply]
Then I think we are done? :-) (And Arno can relax? :-) )

I grabbed some Spanish books from Gutenberg, will see how clean their "Chapter 1"'s are to add to KLA or for your typing apps.

"La cigüeña tocaba cada vez mejor el saxofón y el búho pedía kiwi y queso."
Benjamín pidió una bebida de kiwi y fresa. Noé, sin vergüenza, la más exquisita champaña del menú.
El cadáver de Wamba, rey godo de España, fue exhumado y trasladado en una caja de zinc que pesó un kilo.
El pingüino Wenceslao hizo kilómetros bajo exhaustiva lluvia y frío; añoraba a su querido cachorro. 
El veloz murciélago hindú comía feliz cardillo y kiwi. La cigüeña tocaba el saxofón detrás del palenque de paja. 
El viejo Señor Gómez pedía queso, kiwi y habas, pero le ha tocado un saxofón.
Es extraño mojar queso en la cerveza o probar whisky de garrafa. 
Ese libro explica en su epígrafe las hazañas y aventuras de Don Quijote de la Mancha en Kuwait. 
Extraño pan de col y kiwi se quemó bajo fugaz vaho.
José compró una vieja zampoña en Perú. Excusándose, Sofía tiró su whisky al desagüe de la banqueta.
Jovencillo emponzoñado de whisky, qué mala figurota exhibes.
Jovencillo emponzoñado de whisky: ¡qué figurota exhibe!
La cigüeña gigante bebió ocho copas de whisky, más quince jarras llenas de fría cerveza rubia, y enseguida huyó en un taxi.
La niña, viéndose atrapada en el áspero baúl índigo y sintiendo asfixia, lloró de vergüenza; mientras que la frustrada madre llamaba a su hija diciendo: “¿Dónde estás Waleska?”. 
Queda gazpacho, fibra, látex, jamón, kiwi y viñas.
Quiere la boca exhausta vid, kiwi, piña y fugaz jamón.
Whisky bueno: ¡excitad mi frágil pequeña vejez!

**Reply by binarybottle** [reply]
Whew!!!  On to punctuation today...

**Comment by binarybottle** [comment]
@iandoug -- Zenodo is a great idea. There is also osf.io, which is a free service that I use a lot; it caps files at around 5 gb. We should upload all files and description to either or both service when we are ready, and link from our documentation.

**Reply by iandoug** [reply]
Tried to register at osf but I am NOT doing that {)@({@#($ captcha.

**Comment by binarybottle** [comment]
@iandoug -- are you recommending that we derive counts for the Leipzig (summed across all categories), and also for Gutenberg, and combine these counts as we see fit afterwards?

**Reply by iandoug** [reply]
That is what I did.

In order to get a "combined" bigram score, I multiplied the count for each pair in Gutenberg by (total leipzig)/(total gutenberg), ie scaling the whole corpus to be about the same, then simply added the counts for each bigram together.
I suppose we could do the same for the letters-only counts. Will see how they look.

For your purposes, for the characters, I converted each char's count to a percentage of the total (for Leipzig and Gutenberg separately) , then took the average of the percentages.

So effectively weighted the result as 50% Leipzig and 50% Gutenberg.

**Reply by binarybottle** [reply]
Sounds great!

**Comment by iandoug** [comment]
Attached spreadsheet with bigrams, Leipzig and Gutenberg

1. Letters only
2. lowercase -> uppercase
3. diacritic -> plain

[spanish-bigrams-unicase-letters-leipzig-gutenberg.ods](https://github.com/binarybottle/engram-es/files/6995086/spanish-bigrams-unicase-letters-leipzig-gutenberg.ods)

If you only want first and second last columns, export to csv, close, open csv, delete unwanted columns, save, import ...
I left it as spreadsheet so you could see the formulas.

**Reply by binarybottle** [reply]
You are awesome, @iandoug!  This is exactly what I needed!
Now I can get cranking!!!

**Reply by iandoug** [reply]
Char frequencies still coming.

**Comment by binarybottle** [comment]
I'm working with the assumption that the unicase letter bigram frequency file that I received is final, because this is how I generated the current layout for which I am conducting tests.

**Reply by Lobo-Feroz** [reply]
I don't know if you are referring to the fact that I'm still cleaning the "bad words" list. I don't think the count for very low frequency words would be statistically relevant for the monogram and bigram frequencies that Ian has given us already, for layout optimization purposes. We're scraping the bottom of the barrel for valid but actually pretty rare words here.

But I do think that having a cleaner corpus and probably a good spanish dictionary of valid words can have some other uses, for instance to generate random lists of words for typing training, to validate further spanish corpora, other uses for KLA maybe...

**Reply by binarybottle** [reply]
That makes sense. Thank you! 
I'll continue with what I'm doing...

**Comment by Lobo-Feroz** [comment]
Hey @iandoug 

Gutemberg's corpus distorts symbol count by using them as special characters.

For instance the underscore is used to indicate italicized words.

https://www.gutenberg.org/attic/html_faq.html#can-i-submit-a-html-or-other-format-of-somebody-elses-text
"A plain text file, using extended character sets like ISO-8859 [V.76] or Unicode [V.77] and underscores for italics, can capture all of the author’s intent in almost all cases."

And that's why there are a gajillion underscores in Gutemberg.

This, _obviously_, distorts its frequency, massively. It's true that nowadays underscore has this purpose, many apps like whatsapp, telegram or this very github use underscores for italics, and usually asterisks for **bold**.

The question is: should we entertain this convention in order to create an optimal spanish layout?

**Reply by iandoug** [reply]
Why do Spanish authors put so much in italics? :-)

English books use italics and bold very seldom.

Underscore is also used a lot in code.

I suggest you keep it with dash, and just worry about dash frequency, which is probably also distorted by Gutenberg using -- for &mdash;

**Comment by binarybottle** [comment]
@iandoug -- It would be nice to wrap this up and put the relevant files and description up in osf.io / zenodo.  If you don't have time, I would be happy to do so -- just point me to the most relevant files.  Thanks!

**Reply by iandoug** [reply]
Busy with it. Will revert.

**Reply by iandoug** [reply]
Attached draft for review and comment. Probably still needs a bit of polish.

The "file list" (section 5) is not final yet .

Please ensure I credited everybody correctly. :-)

@NickG13
@binarybottle
@Lobo-Feroz

Thanks, Ian

[Creating-a-corpus-and-chained-bigrams-for-Spanish-keyboard-development-and-evaluation-0.9.0.pdf](https://github.com/binarybottle/engram-es/files/7112340/Creating-a-corpus-and-chained-bigrams-for-Spanish-keyboard-development-and-evaluation-0.9.0.pdf)

**Reply by binarybottle** [reply]
FANTASTIC, @iandoug!  You have done a great job documenting the project, and people will find this very helpful!

**Reply by Lobo-Feroz** [reply]
Awesome work @iandoug !

What a great documentation, both to show that this layout has been thoroughly researched, and as a base for other optimized spanish layouts based on different paradigms. 

And thanks for the credits!

Found a little typo. In "6. Acknowledgements" there's a missing space after the parenthesis and before "Arno" here: 

`Thanks to (alphabetically)Arno Klein`

Also the beginning of the PDF title shows up as:

`Zep Tepi Mathematics 201`

Which looks like a different project altogether. Possibly a field unchanged from the document used as template.

Congratulations again! 👍

**Reply by iandoug** [reply]
Done. Hope all is correct :-)

Report and corpus/analysis files:

https://zenodo.org/record/5501914

all versions: https://doi.org/10.5281/zenodo.5501913

this version: https://doi.org/10.5281/zenodo.5501914

Suspect German will be next, but we need to add deadkey support to KLA ...

Cheers, Ian

**Reply by iandoug** [reply]
> Also the beginning of the PDF title shows up as:
> 
> `Zep Tepi Mathematics 201`
> 
> Which looks like a different project altogether. Possibly a field unchanged from the document used as template.

Yuch. Only saw that after uploading. Only shows up when viewing the PDF.

Uploaded v1.0.1 to fix that. Is indeed from template... I need to get back to Giza :-)

**Comment by binarybottle** [comment]
Congratulations and great work, @iandoug !!!
